### PR TITLE
Split `lineBounds` out into its own method

### PR DESCRIPTION
**Pull request by camdencheek**
Building off #363, this splits `lineBounds` out of `atOffset` and `getLines` to reduce some of the logic duplication and so that we can fetch the boundaries of a single line from a line number rather than just a byte offset. This is another piece of information I will need to enable multiline matches. 

I added a unit test for this new method, and all existing unit tests for `lineBounds` and `getLines` still pass.

### CODE REVIEW

File: contentprovider.go [hunk 1]
```diff
@@ -261,41 +261,49 @@ func (nls newlines) atOffset(offset uint32) (lineNumber, lineStart, lineEnd int)
...
+	} else if endIdx >= len(nls.locs) {
+		end = nls.fileSize
 	}
 
-	return idx + 1, start, end
+	return start, end
 }
 
 // getLines returns a slice of data containing the lines [low, high).
 // low is 1-based and inclusive. high is exclusive.
```
Comment (mpimenov) [inline contentprovider.go]: is high 1-based too? Could you please clarify it here?

File: contentprovider.go [hunk 2]
```diff
@@ -261,41 +261,49 @@ func (nls newlines) atOffset(offset uint32) (lineNumber, lineStart, lineEnd int)
...
 func (nls newlines) getLines(data []byte, low, high int) []byte {
-	// newlines[0] is the start of the 2nd line in data.
-	// So adjust low and high to be based on newLines.
-	low -= 2
-	high -= 2
-	if low >= high || high < 0 || low >= len(nls.locs) || len(nls.locs) == 0 {
```
Comment (mpimenov) [inline contentprovider.go:284]: Please add a test that has empty nls.locs.

File: contentprovider.go [hunk 3]
```diff
@@ -261,41 +261,49 @@ func (nls newlines) atOffset(offset uint32) (lineNumber, lineStart, lineEnd int)
...
+	startIdx := lineNumber - 2
+	endIdx := lineNumber - 1
+
+	if startIdx >= 0 && startIdx < len(nls.locs) {
+		start = nls.locs[startIdx] + 1
+	} else if startIdx < 0 {
+		start = 0
+	} else if startIdx >= len(nls.locs) {
+		start = nls.fileSize
 	}
```
Comment (mpimenov) [inline contentprovider.go:283]: nit: does avoiding duplication help readability here?
```
	if startIdx < 0 {
		start = 0
	} else if startIdx >= len(nls.locs) {
		start = nls.fileSize
	} else {
		start = nls.locs[startIdx] + 1
	}
```

File: contentprovider.go [hunk 4]
```diff
@@ -261,41 +261,49 @@ func (nls newlines) atOffset(offset uint32) (lineNumber, lineStart, lineEnd int)
...
-	start := 0
-	if idx > 0 {
-		start = int(nls.locs[idx-1] + 1)
+	if endIdx >= 0 && endIdx < len(nls.locs) {
+		end = nls.locs[endIdx]
+	} else if endIdx < 0 {
+		end = 0
+	} else if endIdx >= len(nls.locs) {
+		end = nls.fileSize
 	}
```
Comment (mpimenov) [inline contentprovider.go:291]:
```
start = nls.lineStart(lineNumber - 2)
end = nls.lineStart(lineNumber - 1)
```

with the corresponding method, now that we have the newlines type

?

Comment (mpimenov) [reply contentprovider.go:291]: Would also simplify the call sites
(currently
```
lowStart, _ := nls.lineBounds(low)
	_, highEnd := nls.lineBounds(high - 1)
```
)

 and probably remove the need of this method

Comment (camdencheek) [reply contentprovider.go:291]: The problem with this is that, in general, `lineEnd(n) != lineStart(n+1)`. This is because we do not include the newline character in the line bounds. 

We might be able to safely say `lineStart(n) == lineEnd(n-1) + 1`, but only because we clamp offsets returned by `lineStart` to `[0, len(fileContent))`. However, I think it's preferable to keep this behavior a more explicit part of the method call API than to depend on callers to remember to do this. Especially since it takes some convincing to be confident that that is always right in all edge cases (file boundaries, empty lines, no trailing newline, etc.)

File: contentprovider.go (hunk 1 above)
Comment (camdencheek) [reply contentprovider.go]: Done in https://github.com/sourcegraph/zoekt/pull/366/commits/77be96841336944a903a7aeec58c1abf38c392c0

File: contentprovider.go (hunk 2 above)
Comment (camdencheek) [reply contentprovider.go:284]: I updated the `GetLines` test to include a few different cases in https://github.com/sourcegraph/zoekt/pull/366/commits/77be96841336944a903a7aeec58c1abf38c392c0 (the original test only tested a single file).

File: contentprovider.go (hunk 3 above)
Comment (camdencheek) [reply contentprovider.go:283]: Yes! Thanks -- that is significantly easier to understand

**Review by keegancsmith — APPROVED** [review]

File: contentprovider.go [hunk 5]
```diff
@@ -261,41 +261,49 @@ func (nls newlines) atOffset(offset uint32) (lineNumber, lineStart, lineEnd int)
...
+	start, end := nls.lineBounds(idx + 1)
+	return idx + 1, int(start), int(end)
+}
+
+// lineBounds returns the byte offsets of the start and end of the 1-based
+// lineNumber. The end offset is exclusive and will not contain the line-ending
+// newline. If the line number is out of range of the lines in the file, start
+// and end will be clamped to [0,fileSize].
+func (nls newlines) lineBounds(lineNumber int) (start, end uint32) {
+	// newlines[0] is the start of the 2nd line in data.
```
Comment (mpimenov) [inline contentprovider.go:273]: s/newlines/nls.locs/

Comment (mpimenov) [reply contentprovider.go:273]: and this is not true?
start of the 2nd line is ```nls.locs[0] + 1```, isn't it?

File: contentprovider.go (hunk 4 above)
Comment (mpimenov) [reply contentprovider.go:291]: Now I see that they differ by 1. Ok.

**Review by mpimenov — COMMENTED** [review]
Please ping the unfinished reviewers before merging next time.

**Comment by camdencheek** [comment]
> Please ping the unfinished reviewers before merging next time.

Ah! I'm sorry. I saw the approval notification from Keegan and misread it as an approval from you.

**Comment by mpimenov** [comment]
LGTM

File: contentprovider.go (hunk 5 above)
Comment (camdencheek) [reply contentprovider.go:273]: Follow up here: https://github.com/sourcegraph/zoekt/pull/369

**Comment by keegancsmith** [comment]
> Please ping the unfinished reviewers before merging next time.

@mpimenov we generally only block reviews if there are issues with them that we want to take a follow up look in. I think @camdencheek merging was correct, and if he accidently missed something important you can follow up. We shouldn't be blocking reviews until everyone approves. See our guidelines here https://docs.sourcegraph.com/dev/background-information/pull_request_reviews#reviewing-pull-requests

**Comment by mpimenov** [comment]
@keegancsmith I wasn't meaning to block, sorry if it sounded this way. However, I don't think pinging before merging contradicts your link.
Anyway, my understanding is that the dynamics of the follow-up request implies @camdencheek and I understood each other.
